lib/git/lib.rb

```diff
@@ -311,17 +311,32 @@ def full_log_commits(opts = {})
       process_commit_log_data(full_log)
     end
 
-    def revparse(string)
-      return string if string =~ /^[A-Fa-f0-9]{40}$/  # passing in a sha - just no-op it
-      rev = ['head', 'remotes', 'tags'].map do |d|
-        File.join(@git_dir, 'refs', d, string)
-      end.find do |path|
-        File.file?(path)
-      end
-      return File.read(rev).chomp if rev
-      command('rev-parse', string)
+    # Verify and resolve a Git revision to its full SHA
+    #
+    # @see https://git-scm.com/docs/git-rev-parse git-rev-parse
+    # @see https://git-scm.com/docs/git-rev-parse#_specifying_revisions Valid ways to specify revisions
+    # @see https://git-scm.com/docs/git-rev-parse#Documentation/git-rev-parse.txt-emltrefnamegtemegemmasterememheadsmasterememrefsheadsmasterem Ref disambiguation rules
+    #
+    # @example
+    #   lib.rev_parse('HEAD') # => '9b9b31e704c0b85ffdd8d2af2ded85170a5af87d'
+    #   lib.rev_parse('9b9b31e') # => '9b9b31e704c0b85ffdd8d2af2ded85170a5af87d'
+    #
+    # @param revision [String] the revision to resolve
+    #
+    # @return [String] the full commit hash
+    #
+    # @raise [Git::FailedError] if the revision cannot be resolved
+    # @raise [ArgumentError] if the revision is a string starting with a hyphen
+    #
+    def rev_parse(revision)
+      assert_args_are_not_options('rev', revision)
+
+      command('rev-parse', revision)
```

Comment:
Hey @jcouball, thank you for this project.

I wonder if this command could be adjusted like this:
```suggestion
      command('rev-parse', '--revs-only', '--end-of-options', revision, '--')
```
This changes the `exception.result.stderr` to:
```bash
$ git rev-parse --revs-only --end-of-options v3 --
fatal: bad revision 'v3'
```
from:
```bash
$ git rev-parse v3
v3
fatal: ambiguous argument 'v3': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```
Which left me confused and investigating first time I [encountered missing revision name](https://github.com/mpapis/uncov/commit/23c312c29a7594ef8c05ef269b395d398e4c6cf7#diff-9e2a28deb4619667b946a3a8dd073c3fa86839ff5f829b27d339a85bc4167ebbR47-R50).

If you are up for this change I can prepare a PR for it, WDYT?

Reply:
Yes, a PR would be appreciated. 

I do have a couple of caveats. Not sure if they can be answered, but I'll take the PR anyway.

Appending `--` at the end of the command seems to be ambiguous when `--revs-only` is given since (it seems like) git ignores anything given after the `--`. In my tests, the output is the same where I give a path (even if the path does not exist).

What it does change is the error message. With the `--` you get the more helpful error message as you point out.

This makes me wonder:
1. Can this behavior can be counted on long term?
2. Will there come a time when giving `--revs-only` and `--` will be made an error?

Reply:
https://github.com/ruby-git/ruby-git/pull/789 - I think that `--` behavior is so deep at core of git that it will never change.